pkg/api/steve/supportconfigs/generator.go

```diff
@@ -24,33 +25,87 @@ import (
+	configMaps   v1.ConfigMapInterface
+	cspNamespace string
+)
+
 // Handler implements http.Handler - and serves supportconfigs (tar file which contains support relevant information)
 type Handler struct {
-	ConfigMaps           v1.ConfigMapInterface
+	scaledContext        *config.ScaledContext
 	SubjectAccessReviews authv1.SubjectAccessReviewInterface
 	adapterUtil          *cspadapter.ChartUtil
+	billingAdapterUtil   *cspbillingadapter.ChartUtil
 }
 
 // NewHandler creates a handler using the clients defined in scaledContext
 func NewHandler(scaledContext *config.ScaledContext) Handler {
 	return Handler{
-		ConfigMaps:           scaledContext.Core.ConfigMaps(cspadapter.ChartNamespace),
+		scaledContext:        scaledContext,
 		SubjectAccessReviews: scaledContext.K8sClient.AuthorizationV1().SubjectAccessReviews(),
 		adapterUtil:          cspadapter.NewChartUtil(scaledContext.Wrangler.RESTClientGetter),
+		billingAdapterUtil:   cspbillingadapter.NewChartUtil(scaledContext.Wrangler.RESTClientGetter),
 	}
 }
 
+// FIXME: this logic is based on the assumption that Rancher PAYG offering has precedence over
+// BYO offering. Therefore, we'll check to see if CSP billing adapter is installed first.
+// However, there could be a scenario (migration?) where both adapters may co-exist. We'll need
+// to figure out how to handle such scenario in the future.
+// Returns "BYO" for Bring-Your-Own, "PAYG" for Pay-As-You-Go, or "" if neither are found.
+func (h *Handler) getOffering() (string, error) {
+	_, err := h.billingAdapterUtil.GetRelease()
+	if err == nil {
+		return PAYG, nil
+	}
+	if errors.Is(err, cspbillingadapter.ErrNotFound) {
+		_, err := h.adapterUtil.GetRelease()
+		if err == nil {
+			return BYO, nil
+		}
+		if errors.Is(err, cspadapter.ErrNotFound) {
+			return "", nil
+		}
+	}
+	return "", err
+}
+
 // ServerHTTP implements http.Handler - attempts to authenticate/authorize the user. Returns a tar of support information
 // if the user can get the backing configmap in the adapter namespace
 func (h *Handler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	offering, err := h.getOffering()
+	if err != nil {
+		logrus.Errorf("[%s] Error when generating supportconfig %v", logPrefix, err)
+		util.ReturnHTTPError(writer, request, http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+		return
+	}
+	if offering == PAYG {
+		cspNamespace = cspbillingadapter.ChartNamespace
+	} else if offering == BYO {
+		cspNamespace = cspadapter.ChartNamespace
+	} else {
+		// if neither adapter is not installed, return a 501 so the
+		// user knows to install the adapter
+		util.ReturnHTTPError(writer, request, http.StatusNotImplemented, "csp-billing-adapter or csp-adapter must be installed in order to to generate supportconfigs")
+		return
+	}
+	configMaps = h.scaledContext.Core.ConfigMaps(cspNamespace)
+	// FIXME: does it matters if we check for permissions after determining whether the CSP adapter is installed?
```

Comment:
Relevant to this comment, I would recommend only issuing commands based on user input after the request has been authenticated. Doing it before can leak information (for example, in this case an unathenticated user could use this to determine if the adapter is installed), and it doesn't give you any advantage.

Reply:
Without determining the licensing by first looking up the CSP namespace, we can't perform authorization on that namespace, unless we want to perform authorization on both possible namespaces. What's the risk of exposing the fact that CSP Adapter is installed or not installed?

Reply:
> Without determining the licensing by first looking up the CSP namespace, we can't perform authorization on that namespace, unless we want to perform authorization on both possible namespaces.

I think that the authorization process should be as follows:
- We get the CSP release namespace
- If we have a CSP release namespace, authorize against that namespace
- If we don't have the namespace, we authorize against the BYOL namespace. I'm going this route to keep backward compatibility - users who could see the install/uninstall state before should be able to see it now.

This being said, the above has made me a bit uncomfortable at the behavior of this endpoint - since it effectively means that you need two different sets of permissions to use it (one set if the current adapter is used, and a different set if the other adapter is installed). I don't think that this is optimal. 

I'm thinking that it might be better to require users using the new billing adapter to specify a parameter in the get request (something like "?usePAYG=true") so that we can specify exactly which adapter's supportconfig they want. I think this would also help with the override case that you labeled as a FIXME in the code, since a request can now only target one adapter's supportconfig. What do you think of that option?

> What's the risk of exposing the fact that CSP Adapter is installed or not installed?

Hard to quantify the exact risk, but this is information that the user doesn't have a right to have, so we don't give it to them, ever. Making small exceptions like this leads to larger security risks, so it's key to maintain best practices here.

Reply:
I like the benefit of using `usePAYG=true`. One potential drawback is that we may need to have two different versions (branding) of dashboard.

Reply:
We don't want two versions of the dashboard, if we can't check for the presence of the different helm charts could we set what type of environment we're in via a config or environment variable/setting at install time? Theres bound to be a way to implement this without forking the dashboard, which is definitely a no-go.

Reply:
I assume the dashboard could include a parameter based on the check for the adapter going on at https://github.com/rancher/dashboard/pull/9092/files#diff-556137f68d06f23031944f124827daa23f6f1cb5c7c6fe4ade07260010874f9eR11  , but that raises the question of whether the same concern raised here applies to that as well... if not, then the UI page triggering the supportconfig generation should be able to, in theory, know whether its PAYG or not

Reply:
I think we can do this in the UI, assuming it is the only consumer of the support config endpoint. UI will not show the generate support config button unless one of the CSP adapters is detected. But if we want to be explicit about which licensing UI is running under, then we'll need some mechanism to tell the UI instead of helm chart detection. Hence my previuse concern with different UI branding.

Reply:
> but that raises the question of whether the same concern raised here applies to that as well..

I don't think that it does. The UI runs on the client, so if the user doesn't have permissions to see the installed charts they just won't see the button to generate the supportconfig. Meaning that the UI can't leak data to the user - it only knows what the user knows, it has to use the user's credentials to retrieve the information. 

> One potential drawback is that we may need to have two different versions (branding) of dashboard.

I also don't think that this is necessary. I'm thinking that a simple "if csp-adapter, no param, if billing adapter, param, if I can't tell, no param" would work fine, no different branding needed.

> I think we can do this in the UI, assuming it is the only consumer of the support config endpoint. 

As far as I know, the UI is the only consumer. Even if there is a script-level consumer, since leaving out the parameter keeps the current behavior I think this solution would be acceptable.

Reply:
Here's the corresponding changes for dashboard https://github.com/rancher/dashboard/pull/9092

Reply:
Resolving, new changes authorize before returning to the user.